AMS 2 Changes + Multi-Sim Track Info HUD

RiddleTime opened the pull request:
Hi @ConnorMolz !

I saw you've forked, I can gave you a few tips here and there and this PR should allow you to also ask questions.

ConnorMolz committed:
Create TrackInfoOverlay.cs in common

(Convert from ACC Solution)

ConnorMolz committed:
Add Flags to sessionData and TrackInfoOverlay.cs with Ams2Mapper.cs

ConnorMolz committed:
Remove grip, because it's not implemented in Ams2 Shared Memory

ConnorMolz committed:
Change variable to prevent data inconsistency

RiddleTime commented on Race Element.Data/Games/Automobilista2/DataMapper/Ams2Mapper.cs:
```diff
@@ -107,6 +107,60 @@ public static void ToLocalSession(Shared shared, SessionData session)
         }
+        
+        // Flag Mapper
+        switch ((Constants.RaceFlags)shared.mHighestFlagColour)
```
Give Switch Expressions a shot: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression

It's a more compact syntax to write a switch case where you for example map between 2 enums.

(example taken straight from the above link).
```csharp
public static class SwitchExample
{
    public enum Direction
    {
        Up,
        Down,
        Right,
        Left
    }

    public enum Orientation
    {
        North,
        South,
        East,
        West
    }

    public static Orientation ToOrientation(Direction direction) => direction switch
    {
        Direction.Up    => Orientation.North,
        Direction.Right => Orientation.East,
        Direction.Down  => Orientation.South,
        Direction.Left  => Orientation.West,
        _ => throw new ArgumentOutOfRangeException(nameof(direction), $"Not expected direction value: {direction}"),
    };

    public static void Main()
    {
        var direction = Direction.Right;
        Console.WriteLine($"Map view direction is {direction}");
        Console.WriteLine($"Cardinal orientation is {ToOrientation(direction)}");
        // Output:
        // Map view direction is Right
        // Cardinal orientation is East
    }
}
```

ConnorMolz replied:
> Give Switch Expressions a shot: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression
> 
> It's a more compact syntax to write a switch case where you for example map between 2 enums.
> 
> (example taken straight from the above link).
> 
> ```cs
> public static class SwitchExample
> {
>     public enum Direction
>     {
>         Up,
>         Down,
>         Right,
>         Left
>     }
> 
>     public enum Orientation
>     {
>         North,
>         South,
>         East,
>         West
>     }
> 
>     public static Orientation ToOrientation(Direction direction) => direction switch
>     {
>         Direction.Up    => Orientation.North,
>         Direction.Right => Orientation.East,
>         Direction.Down  => Orientation.South,
>         Direction.Left  => Orientation.West,
>         _ => throw new ArgumentOutOfRangeException(nameof(direction), $"Not expected direction value: {direction}"),
>     };
> 
>     public static void Main()
>     {
>         var direction = Direction.Right;
>         Console.WriteLine($"Map view direction is {direction}");
>         Console.WriteLine($"Cardinal orientation is {ToOrientation(direction)}");
>         // Output:
>         // Map view direction is Right
>         // Cardinal orientation is East
>     }
> }
> ```

Is changed like you wished

ConnorMolz replied:
I also implement it at Phase getting and Session type getting

ConnorMolz commented:
Hi,

Thank you!

I have some Problems with AMS2...

I dont know but i guess something wirh the datamapper is broken, I will investigating in today and tomorrow

RiddleTime commented:
> Hi,
> 
> Thank you!
> 
> I have some Problems with AMS2...
> 
> I dont know but i guess something wirh the datamapper is broken, I will investigating in today and tomorrow

I personally don't own AMS2, @iFuSiiOnzZ who has implemented the AMS2 data mapping should do.

You're also welcome to join the race element discord if that's easier.

ConnorMolz commented:
The Problem is, that i dont got the data in the HUD...

I implemented it the same way like the current gear

RiddleTime commented:
> The Problem is, that i dont got the data in the HUD...
> 
> I implemented it the same way like the current gear.
> 
> I also got no data for the Lap Data Bar

did you go to options > System> enable the Telemetry?

ConnorMolz commented:
This is not the problem i double checked it and tried it with Crew Chief and it worked.

As example the delta does not work but the rpm and current gear are working

iFuSiiOnzZ commented:
Ah, ok. There is a problem with the multi sim. Currently, the data structures are not generic, the ACC ones are used. The AMS2 mapper translates from AMS2 to ACC data structures, but since it is not a 1 to 1 translation, it is very difficult to have exactly the same HUD.

PS1: We should put pressure on @RiddleTime to have generic data structure haha
PS2: @ConnorMolz I had a look at the shared memory and I am not seeing AMS2 reporting delta

ConnorMolz commented:
Ok, I only looked on the data I needed for the trackInfo so I saw at testing that  it does not work, but I dont investigated it until now

ConnorMolz committed:
Merge branch 'RiddleTime:dev' into dev

ConnorMolz committed:
Update Data Model for flags:
- remove duplicate finish flag
- and Black with Orange circle (damage) and Penalty flage (black and white)

ConnorMolz committed:
Fix the AMS2 flag mapper like wished in comment

ConnorMolz committed:
Merge remote-tracking branch 'origin/dev' into dev

ConnorMolz committed:
Add ToSession Type

ConnorMolz committed:
Add ToSessionPhase to the AMS2 Data mapper

ConnorMolz commented:
> discord

I am already on the Discord. I changed my name to the GitHub name

ConnorMolz commented:
> > Hi,
> > Thank you!
> > I have some Problems with AMS2...
> > I dont know but i guess something wirh the datamapper is broken, I will investigating in today and tomorrow
> 
> I personally don't own AMS2, @iFuSiiOnzZ who has implemented the AMS2 data mapping should do.
> 
> You're also welcome to join the race element discord if that's easier.

I am already on the Discord. I changed my name to the GitHub name

RiddleTime commented on Race Element.Data/Common/SimulatorData/SessionData.cs:
```diff
@@ -116,3 +118,18 @@ public enum Status : int
     PAUSE,
 }
+
+public enum CurrentFlag : int
```
An idea could be to add the [Flags] Attribute to this enum. Just like the Game Enum. This will allow for a green flag + meatball(black-orange) at the same time.

```sharp
CurrentFlag twoFlags = CurrentFlag.Green | CurrentFlag.Damage;
```

```csharp
[Flags]
public enum Game : int
{
    Any = (1 << 0),
    AssettoCorsa1 = (1 << 1),
    AssettoCorsaCompetizione = (1 << 2),
    iRacing = (1 << 3),
    RaceRoom = (1 << 4),
    Automobilista2 = (1 << 5),
    EuroTruckSimulator2 = (1 << 6),
    AmericanTruckSimulator = (1 << 7),
    //rFactor2,
    // LMU,
}
```

ConnorMolz replied:
So for AMS 2 is it not possible because, AMS2 is reporting only the highest flag so in this case it's the damageflag 


Data from AMS2 Constants
```public enum RaceFlags // (Type#5) Flag Colours (to be used with 'mHighestFlagColour')
    {
        FLAG_COLOUR_NONE = 0,               // Not used for actual flags, only for some query functions
        FLAG_COLOUR_GREEN,                  // End of danger zone, or race started
        FLAG_COLOUR_BLUE,                   // Faster car wants to overtake the participant
        FLAG_COLOUR_WHITE_SLOW_CAR,         // Slow car in area
        FLAG_COLOUR_WHITE_FINAL_LAP,        // Final Lap
        FLAG_COLOUR_RED,                    // Huge collisions where one or more cars become wrecked and block the track
        FLAG_COLOUR_YELLOW,                 // Danger on the racing surface itself
        FLAG_COLOUR_DOUBLE_YELLOW,          // Danger that wholly or partly blocks the racing surface
        FLAG_COLOUR_BLACK_AND_WHITE,        // Unsportsmanlike conduct
        FLAG_COLOUR_BLACK_ORANGE_CIRCLE,    // Mechanical Failure
        FLAG_COLOUR_BLACK,                  // Participant disqualified
        FLAG_COLOUR_CHEQUERED,              // Chequered flag
        FLAG_COLOUR_MAX
    }```

RiddleTime replied:
The bitwise flags are very useful and effecient to crosscheck against multiple enum values. 

```csharp
CurrentFlag yellowFlags = CurrentFlag.Yellow | CurrentFlag.DoubleYellow;
CurrentFlag simulatorFlag = SimdataProvider.....;
// check if the current simulator flag is any of the yellow flags...
if (yellowFlags.HasFlag(simulatorFlag)) { ..... }
```

```csharp
[Flags]
public enum CurrentFlag : int
{
    None = 1 << 0,
    Green = 1 << 1,
    Blue = 1 << 2,
    Yellow = 1 << 3,
    DoubleYellow = 1 << 4,
    Red = 1 << 5,
    Black = 1 << 6,
    White = 1 << 7,
    Checkered = 1 << 8,
    Damage = 1 << 9,
    Penalty = 1 << 10,
    Max = 1 << 11,
}
```

https://learn.microsoft.com/en-us/dotnet/api/system.enum.hasflag?view=net-9.0

ConnorMolz committed:
Merge remote-tracking branch 'upstream/dev' into dev

ConnorMolz committed:
Fix Value error on overlay

ConnorMolz commented:
We currently get data with the last commit, but the wind direction is currently a bit buggy also the wind direction

So still WIP but it is going forward

ConnorMolz committed:
Add double Yellow to the possible flags

ConnorMolz committed:
Implement Wind direction and Wind speed

ConnorMolz committed:
Fix field sizes and improve wind direction performance

ConnorMolz committed:
Fix field sizes and improve wind direction performance

ConnorMolz committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	Race Element.HUD.Common/Overlays/Driving/TrackInfo/TrackInfoOverlay.cs

ConnorMolz commented:
Ready for review!

Is running on my AMS2 without problems.

@iFuSiiOnzZ

ConnorMolz committed:
Resize value field because if the wind speed is greater than 10 the h of km/h is disappearing

ConnorMolz committed:
Updated description

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/TrackInfo/TrackInfoOverlay.cs:
```diff
@@ -0,0 +1,235 @@
+        _windDirectionLabel.Draw(g, "Direction", this.Scale);
+        _windDirectionValue.Draw(g, $"{windDirection}", this.Scale);
+        
+        string windSpeed = (SimDataProvider.Session.Weather.AirVelocity * 3.6).ToString("F2");
```
The air speed in the multi-sim data should already be Kilometers per hour.

ConnorMolz replied:
Nop not in AMS2 found this value in another repo

ConnorMolz replied:
But I will chage it in the mapper later

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/TrackInfo/TrackInfoOverlay.cs:
```diff
@@ -0,0 +1,235 @@
+        }
+        
+        string windDirection = RadianToDirection(SimDataProvider.Session.Weather.AirDirection);
+        _windDirectionLabel.Draw(g, "Direction", this.Scale);
```
It would be better to add the wind direction to the already existing windspeed value.

```csharp
string windDirection = RadianToDirection(SimDataProvider.Session.Weather.AirDirection);
string windSpeed = (SimDataProvider.Session.Weather.AirVelocity).ToString("F2");
_windLabel.Draw(g, "Wind", this.Scale);
_windValue.Draw(g, $"{windSpeed} > {windDirection}", this.Scale);
```

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/TrackInfo/TrackInfoOverlay.cs:
```diff
@@ -0,0 +1,235 @@
+        
+        int sectorIndex = (int)(radian / (twoPi / 8));
+
+        switch (sectorIndex)
```
```csharp
  return sectorIndex switch
  {
      0 => "E",
      1 => "NE",
      2 => "N",
      3 => "NW",
      4 => "W",
      5 => "SW",
      6 => "S",
      _ => "SE",
  };
```

RiddleTime commented:
> Ready for review!
> 
> Is running on my AMS2 without problems.
> 
> @iFuSiiOnzZ @RiddleTime
> 
> Edit: I leet a friend test it and he showed me that it work

The only thing that would need a proper check is whether the unscaled Width and Height of the Track Info HUD can be made smalller (through the review comments perhaps).

RiddleTime reviewed (commented):
In general, I try to keep the HUDs as minimalistic/small as possible.

Few small code style changes here and there to improve upon.

I think after these changes it would be good for release. Will have a look later if we can add some configurables or perhaps nicer drawing of the wind direction (arrows).

ConnorMolz committed:
Resize the Overlay,
Put the "magic value" into the Mapper,
Improve the switch at Wind direction,
Merge the Wind direction and wind speed row together

ConnorMolz commented:
I don't understand how do you mean that with the bitwise flags ^^'
The other points are implemented

RiddleTime commented on Race Element.Data/Games/Automobilista2/DataMapper/Mapper.cs:
```diff
@@ -14,7 +14,8 @@ public static void ToLocalSession(Shared shared, SessionData session)
         session.Weather.AirDirection = (float)Math.Atan2(shared.mWindDirectionY, shared.mWindDirectionX);
         session.Weather.AirTemperature = shared.mAmbientTemperature;
-        session.Weather.AirVelocity = shared.mWindSpeed;
+        session.Weather.AirVelocity = Convert.ToSingle(shared.mWindSpeed + 3.6);
```
I'll merge and fix. But multiply!

RiddleTime commented:
> I don't understand how do you mean that with the bitwise flags ^^' The other points are implemented

Say we want to check in HUD code whether there are any warning flags, we could write an if clause with all the warning flags. Or simply store the warning flags as a variable (stores all the "active bits" within 1 variable).
```csharp
// get the current flag.
CurrentFlag simFlag = SimdataProvider.....;

// check for a warning flag, but check individually.
if (simFlag == CurrentFlag.Yellow || simFlag == CurrentFlag.DoubleYellow || simflag == CurrentFlag.Red) { ... }

// instead with the bit fields ( [Flag] attribute ) we can store the warning flags as a predefined variable.
CurrentFlag warningFlags = CurrentFlag.Yellow | CurrentFlag.DoubleYellow | CurrentFlag.Red;
if (warningFlags.hasFlag(simFlag) { ... }

```
